File: clients/kotlin/Fylo.kt
// FYLO Android client — a local-first document store for Kotlin.
//
// A phone can't spawn the `fylo` binary, so this hosts FYLO's local-first web
// engine (`fylo.mjs`) in a headless android.webkit.WebView: all reads and
// writes hit an on-device OPFS store — fully offline, no backend. It mirrors
// the browser client — same engine, native API.
//
//   val db = Fylo.open(context)
//   db.createCollection("users")
//   val id = db.putData("users", mapOf("name" to "Ada", "role" to "admin")) as String
//   val doc = db.getLatest("users", id)
//   val admins = db.findDocs("users", mapOf("\$ops" to listOf(mapOf("role" to mapOf("\$eq" to "admin")))))
//
// Bundle three files under the app's assets at `assets/fylo/`: this engine's
// `fylo.mjs` (from a FYLO release), plus `host.html` and `bridge.js` from
// clients/mobile/. They are served over an https origin (a secure context,
// which OPFS requires — a file:// origin does not get OPFS) via request
// interception; no network is used for the assets.
//
// Platform only: android.webkit + org.json (both in the SDK) and
// kotlinx.coroutines (standard on Android). No third-party HTTP/JSON libs.
// Requires a Chromium WebView with OPFS (Android System WebView / Chrome 108+).

import android.annotation.SuppressLint
import android.content.Context
import android.graphics.Bitmap
import android.net.Uri
import android.os.Handler
import android.os.Looper
import android.webkit.JavascriptInterface
import android.webkit.RenderProcessGoneDetail
import android.webkit.WebResourceRequest
import android.webkit.WebResourceError
import android.webkit.WebResourceResponse
import android.webkit.WebView
import android.webkit.WebViewClient
import kotlinx.coroutines.CancellationException
import kotlinx.coroutines.CompletableDeferred
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import kotlinx.coroutines.withTimeoutOrNull
import org.json.JSONArray
import org.json.JSONObject
import java.io.ByteArrayInputStream
import java.util.concurrent.ConcurrentHashMap
import java.util.concurrent.atomic.AtomicBoolean
import java.util.concurrent.atomic.AtomicInteger

class Fylo private constructor(
    private val webView: WebView,
    private val rpcTimeoutMillis: Long,
) {
    private val nextId = AtomicInteger(1)
    private val pending = ConcurrentHashMap<Int, CompletableDeferred<Any?>>()
    private val ready = CompletableDeferred<Unit>()
    private val closed = AtomicBoolean(false)
    private val disposed = AtomicBoolean(false)

    /** Send one RPC to the engine and await its result (call off the main thread). */
    suspend fun request(method: String, args: Map<String, Any?> = emptyMap()): Any? {
        if (closed.get()) throw FyloException("FYLO client is closed; reopen it before retrying")
        val id = nextId.getAndIncrement()
        val payload = JSONObject()
            .put("id", id)
            .put("method", method)
            .put("args", toJsonObject(args))
        val payloadBytes = payload.toString().toByteArray(Charsets.UTF_8)
        if (payloadBytes.size > MAX_BRIDGE_REQUEST_BYTES) {
            throw FyloException("FYLO request exceeds the 6 MiB native bridge limit")
        }
        val base64 = android.util.Base64.encodeToString(
            payloadBytes,
            android.util.Base64.NO_WRAP,
        )
        val deferred = CompletableDeferred<Any?>()
        synchronized(pending) {
            if (closed.get()) {
                throw FyloException("FYLO client is closed; reopen it before retrying")
            }
            if (pending.size >= MAX_PENDING_REQUESTS) {
                throw FyloException("too many pending FYLO requests")
            }
            pending[id] = deferred
        }
        try {
            val completed = withTimeoutOrNull(rpcTimeoutMillis) {
                withContext(Dispatchers.Main) {
                    // base64 body is safe inside the JS string literal.
                    webView.evaluateJavascript("window.__fyloDispatch(\"$base64\")", null)
                }
                listOf(deferred.await())
            }
            if (completed != null) return completed.single()
            val error = FyloException(
                "FYLO request '$method' timed out after ${rpcTimeoutMillis}ms; verify the WebView is responsive",
            )
            if (removePending(id, deferred)) deferred.completeExceptionally(error)
            throw error
        } catch (error: CancellationException) {
            if (removePending(id, deferred)) deferred.cancel(error)
            throw error
        } finally {
            if (removePending(id, deferred) && !deferred.isCompleted) {
                deferred.cancel(CancellationException("FYLO request was cancelled"))
            }
        }
    }

    // JS -> native bridge target (runs on a WebView binder thread).
    private inner class Bridge {
        @JavascriptInterface
        fun onMessage(text: String) {
            val recoveredId = recoverReplyId(text)
            if (text.toByteArray(Charsets.UTF_8).size > MAX_BRIDGE_RESPONSE_BYTES) {
                recoveredId?.let {
                    failCorrelated(it, "FYLO bridge response exceeds the 6 MiB limit")
                }
                return
            }
            val reply = try {
                JSONObject(text)
            } catch (_: Exception) {
                recoveredId?.let { failCorrelated(it, "invalid FYLO bridge response") }
                return
            }
            val id = reply.optInt("id", -1)
            if (id == 0) { // bridge-ready signal
                ready.complete(Unit)
                return
            }
            val deferred = pending.remove(id) ?: return
            if (reply.optBoolean("ok", false)) {
                deferred.complete(if (reply.isNull("result")) null else reply.opt("result"))
            } else {
                val message = reply.optJSONObject("error")?.optString("message") ?: "fylo error"
                deferred.completeExceptionally(FyloException(message))
            }
        }

        private fun failCorrelated(id: Int, message: String) {
            val error = FyloException(message)
            if (id == 0) {
                ready.completeExceptionally(error)
                return
            }
            pending.remove(id)?.completeExceptionally(error)
        }
    }

    // --- Collections ---
    // `kind` is accepted for cross-client parity; the on-device engine stores documents untyped.
    suspend fun createCollection(collection: String, kind: String = "document"): Any? =
        request("createCollection", mapOf("collection" to collection))
    suspend fun dropCollection(collection: String): Any? =
        request("dropCollection", mapOf("collection" to collection))
    suspend fun inspectCollection(collection: String): Any? =
        request("inspectCollection", mapOf("collection" to collection))
    suspend fun rebuildCollection(collection: String): Any? =
        request("rebuildCollection", mapOf("collection" to collection))

    // --- Documents (object args are native Maps/Lists) ---
    suspend fun putData(collection: String, data: Map<String, Any?>): Any? =
        request("putData", mapOf("collection" to collection, "data" to data))
    suspend fun getDoc(collection: String, id: String): Any? =
        request("getDoc", mapOf("collection" to collection, "id" to id))
    suspend fun getMeta(collection: String, id: String): Any? =
        request("getMeta", mapOf("collection" to collection, "id" to id))
    suspend fun setMeta(collection: String, id: String, meta: Map<String, Any?>): Any? =
        request("setMeta", mapOf("collection" to collection, "id" to id, "meta" to meta))
    suspend fun getLatest(collection: String, id: String): Any? =
        request("getLatest", mapOf("collection" to collection, "id" to id))
    suspend fun patchDoc(collection: String, id: String, newDoc: Map<String, Any?>): Any? =
        request("patchDoc", mapOf("collection" to collection, "id" to id, "newDoc" to newDoc))
    suspend fun delDoc(collection: String, id: String): Any? =
        request("delDoc", mapOf("collection" to collection, "id" to id))
    suspend fun restoreDoc(collection: String, id: String): Any? =
        request("restoreDoc", mapOf("collection" to collection, "id" to id))

    // --- Query ---
    suspend fun findDocs(collection: String, query: Map<String, Any?>): Any? =
        request("findDocs", mapOf("collection" to collection, "query" to query))

    /**
     * Run raw SQL against the local store. Native interpolation is verbatim —
     * escape/validate untrusted input yourself.
     */
    suspend fun sql(statement: String): Any? =
        request("executeSQL", mapOf("sql" to statement))

    /**
     * Collection-scoped facade with short method names, so
     * `db.collection("users").put(data)` reads like the browser client.
     */
    fun collection(name: String): Collection = Collection(name)

    /** A collection-scoped view; methods drop the leading collection argument. */
    inner class Collection(private val name: String) {
        suspend fun create(kind: String = "document") = createCollection(name, kind)
        suspend fun drop() = dropCollection(name)
        suspend fun inspect() = inspectCollection(name)
        suspend fun rebuild() = rebuildCollection(name)
        suspend fun put(data: Map<String, Any?>) = putData(name, data)
        suspend fun get(id: String) = getDoc(name, id)
        suspend fun getMeta(id: String) = getMeta(name, id)
        suspend fun setMeta(id: String, meta: Map<String, Any?>) = setMeta(name, id, meta)
        suspend fun latest(id: String) = getLatest(name, id)
        suspend fun patch(id: String, newDoc: Map<String, Any?>) = patchDoc(name, id, newDoc)
        suspend fun delete(id: String) = delDoc(name, id)
        suspend fun restore(id: String) = restoreDoc(name, id)
        suspend fun find(query: Map<String, Any?>) = findDocs(name, query)
    }

    /** Release the WebView. Call from the main thread. */
    fun close() {
        failBridge(FyloException("FYLO client is closed; reopen it before retrying"))
        disposeWebView()
    }

    private fun failBridge(error: FyloException) {
        if (!closed.compareAndSet(false, true)) return
        ready.completeExceptionally(error)
        val requests = synchronized(pending) {
            val snapshot = pending.values.toList()
            pending.clear()
            snapshot
        }
        requests.forEach { it.completeExceptionally(error) }
    }

    private fun removePending(id: Int, expected: CompletableDeferred<Any?>): Boolean =
        synchronized(pending) {
            if (pending[id] !== expected) return@synchronized false
            pending.remove(id)
            true
        }

    private fun disposeWebView() {
        if (!disposed.compareAndSet(false, true)) return
        val dispose = Runnable {
            webView.stopLoading()
            webView.removeJavascriptInterface("__fyloNative")
            webView.destroy()
        }
        if (Looper.myLooper() == Looper.getMainLooper()) dispose.run()
        else Handler(Looper.getMainLooper()).post(dispose)
    }

    class FyloException(message: String) : RuntimeException(message)

    companion object {
        private const val HOST = "fylo.localhost"
        private const val MAX_BRIDGE_REQUEST_BYTES = 6 * 1024 * 1024
        private const val MAX_BRIDGE_RESPONSE_BYTES = 6 * 1024 * 1024
        private const val MAX_PENDING_REQUESTS = 256
        private const val DEFAULT_RPC_TIMEOUT_MILLIS = 30_000L
        private const val MAX_RPC_TIMEOUT_MILLIS = 5 * 60_000L
        private val ALLOWED_ASSET_PATHS = setOf("/host.html", "/bridge.js", "/fylo.mjs")

        /** Boot the engine. Suspends until the local store is ready. */
        @SuppressLint("SetJavaScriptEnabled")
        suspend fun open(
            context: Context,
            rpcTimeoutMillis: Long = DEFAULT_RPC_TIMEOUT_MILLIS,
        ): Fylo {
            require(rpcTimeoutMillis in 1..MAX_RPC_TIMEOUT_MILLIS) {
                "FYLO RPC timeout must be between 1ms and 300000ms"
            }
            var candidate: Fylo? = null
            var initializationSucceeded = false
            var initializationFailure = FyloException("FYLO WebView initialization failed")
            try {
                val client = withTimeoutOrNull(rpcTimeoutMillis) {
                    val instance = withContext(Dispatchers.Main) {
                    val created = Fylo(WebView(context), rpcTimeoutMillis)
                    candidate = created
                    created.webView.settings.javaScriptEnabled = true
                    created.webView.settings.domStorageEnabled = true
                    created.webView.addJavascriptInterface(created.Bridge(), "__fyloNative")
                    // Serve the bundled assets over an https origin so OPFS is granted;
                    // intercept every request to that origin — no network is used.
                    created.webView.webViewClient = object : WebViewClient() {
                    override fun shouldOverrideUrlLoading(view: WebView, req: WebResourceRequest): Boolean {
                        val allowed = isAllowedUrl(req.url, documentOnly = true)
                        if (!allowed) {
                            created.failBridge(
                                FyloException("FYLO WebView blocked an unexpected navigation"),
                            )
                            created.disposeWebView()
                        } else if (created.ready.isCompleted) {
                            created.failBridge(
                                FyloException("FYLO WebView unexpectedly navigated after becoming ready"),
                            )
                            created.disposeWebView()
                            return true
                        }
                        return !allowed
                    }

                    override fun onPageStarted(view: WebView, url: String, favicon: Bitmap?) {
                        if (created.ready.isCompleted && !created.closed.get()) {
                            created.failBridge(
                                FyloException("FYLO WebView unexpectedly reloaded after becoming ready"),
                            )
                            created.disposeWebView()
                        }
                    }

                    override fun onReceivedError(
                        view: WebView,
                        req: WebResourceRequest,
                        error: WebResourceError,
                    ) {
                        if (req.isForMainFrame || isAllowedUrl(req.url)) {
                            created.failBridge(
                                FyloException("FYLO WebView failed to load: ${error.description}"),
                            )
                            created.disposeWebView()
                        }
                    }

                    override fun onReceivedHttpError(
                        view: WebView,
                        req: WebResourceRequest,
                        response: WebResourceResponse,
                    ) {
                        if (req.isForMainFrame || isAllowedUrl(req.url)) {
                            created.failBridge(
                                FyloException("FYLO WebView load failed with HTTP ${response.statusCode}"),
                            )
                            created.disposeWebView()
                        }
                    }

                    override fun onRenderProcessGone(
                        view: WebView,
                        detail: RenderProcessGoneDetail,
                    ): Boolean {
                        created.failBridge(
                            FyloException("FYLO WebView render process terminated; reopen the client"),
                        )
                        created.disposeWebView()
                        return true
                    }

                    override fun shouldInterceptRequest(
                        view: WebView,
                        req: WebResourceRequest,
                    ): WebResourceResponse {
                        if (req.method != "GET" || !isAllowedUrl(req.url)) return deniedResponse()
                        val name = req.url.lastPathSegment!!
                        return try {
                            val bytes = context.assets.open("fylo/$name").readBytes()
                            WebResourceResponse(mimeFor(name), "utf-8", ByteArrayInputStream(bytes))
                        } catch (_: Exception) {
                            WebResourceResponse("text/plain", "utf-8", 404, "Not Found", null, null)
                        }
                        }
                    }
                    created.webView.loadUrl("https://$HOST/host.html")
                    created
                }
                    instance.ready.await()
                    instance
                }
                if (client == null) {
                    throw FyloException("FYLO WebView did not become ready within ${rpcTimeoutMillis}ms")
                }
                client.request("open", mapOf("config" to JSONObject()))
                initializationSucceeded = true
                return client
            } catch (error: CancellationException) {
                initializationFailure =
                    FyloException("FYLO WebView initialization was cancelled; reopen the client")
                throw error // Preserve structured coroutine cancellation.
            } catch (error: Throwable) {
                initializationFailure = FyloException(
                    "FYLO WebView initialization failed: ${error.message ?: error::class.java.simpleName}",
                )
                throw error
            } finally {
                if (!initializationSucceeded) {
                    candidate?.failBridge(initializationFailure)
                    candidate?.disposeWebView()
                }
            }
        }

        // Recursively convert native containers to org.json types so nested
        // objects/arrays serialize correctly (JSONObject(Map) does not deep-wrap).
        private fun toJsonObject(map: Map<*, *>): JSONObject {
            val obj = JSONObject()
            for ((key, value) in map) obj.put(key.toString(), toJsonValue(value))
            return obj
        }

        private fun toJsonValue(value: Any?): Any {
            return when (value) {
                null -> JSONObject.NULL
                is Map<*, *> -> toJsonObject(value)
                is Iterable<*> -> JSONArray().apply { value.forEach { put(toJsonValue(it)) } }
                is Array<*> -> JSONArray().apply { value.forEach { put(toJsonValue(it)) } }
                else -> value
            }
        }

        private fun isAllowedUrl(url: Uri, documentOnly: Boolean = false): Boolean {
            val path = url.encodedPath ?: return false
            return url.scheme == "https" &&
                url.host == HOST &&
                url.userInfo == null &&
                url.port == -1 &&
                url.query == null &&
                url.fragment == null &&
                path in ALLOWED_ASSET_PATHS &&
                (!documentOnly || path == "/host.html")
        }

        private fun deniedResponse(): WebResourceResponse =
            WebResourceResponse("text/plain", "utf-8", 403, "Forbidden", emptyMap(), null)

        private fun recoverReplyId(text: String): Int? {
            val match = Regex("""^\s*\{\s*"id"\s*:\s*([0-9]{1,15})(?=\s*[,}])""")
                .find(text.take(1024))
            return match?.groupValues?.get(1)?.toIntOrNull()?.takeIf { it >= 0 }
        }

        private fun mimeFor(name: String): String = when {
            name.endsWith(".html") -> "text/html"
            name.endsWith(".js") || name.endsWith(".mjs") -> "text/javascript"
            name.endsWith(".json") -> "application/json"
            else -> "application/octet-stream"
        }
    }
}

File: clients/swift/Fylo.swift
// FYLO iOS client — a local-first document store for Swift.
//
// Unlike the server/CLI shims, a phone can't spawn the `fylo` binary. Instead
// this hosts FYLO's local-first web engine (`fylo.mjs`) in a headless WKWebView:
// all reads and writes hit an on-device OPFS store — fully offline, no backend.
// It mirrors the browser client — same engine, native API.
//
//   let db = try await Fylo()
//   try await db.createCollection("users")
//   let id = try await db.putData("users", ["name": "Ada", "role": "admin"]) as! String
//   let doc = try await db.getLatest("users", id)
//   let admins = try await db.findDocs("users", ["$ops": [["role": ["$eq": "admin"]]]])
//
// Bundle three files as app resources: this engine's `fylo.mjs` (from a FYLO
// release), plus `host.html` and `bridge.js` from clients/mobile/. They are
// served over a custom `fylo-app://` scheme — a secure origin, which OPFS
// requires (a file:// origin does not get OPFS).
//
// Requires iOS 14+ / macOS 11+ (WebKit). The WebView is never shown.

import Foundation
import WebKit

enum FyloError: Error, CustomStringConvertible {
    case failed(String)
    var description: String {
        switch self { case .failed(let message): return message }
    }
}

@MainActor
final class Fylo: NSObject {
    private static let maxBridgeBytes = 6 * 1024 * 1024
    private static let maxPendingRequests = 256
    private static let maxRPCTimeout: TimeInterval = 5 * 60
    private struct PendingRequest {
        let continuation: CheckedContinuation<Any?, Error>
        let timeout: DispatchWorkItem
    }
    private let webView: WKWebView
    private let scheme: String
    private let rpcTimeout: TimeInterval
    private var nextId = 1
    private var pending: [Int: PendingRequest] = [:]
    private var readyContinuation: CheckedContinuation<Void, Error>?
    private var readyTimeout: DispatchWorkItem?
    private var isReady = false
    private var isClosed = false

    /// Boot the engine. Resolves once the local store is ready. `bundle` holds
    /// the three engine assets. Custom scheme host defaults to `fylo-app`.
    init(
        bundle: Bundle = .main,
        scheme: String = "fylo-app",
        rpcTimeout: TimeInterval = 30
    ) async throws {
        guard scheme.range(of: "^[a-z][a-z0-9+.-]{0,31}$", options: .regularExpression) != nil,
              !["http", "https", "file", "data", "javascript"].contains(scheme)
        else { throw FyloError.failed("invalid FYLO WebView scheme") }
        guard rpcTimeout.isFinite, rpcTimeout > 0, rpcTimeout <= Self.maxRPCTimeout else {
            throw FyloError.failed("FYLO RPC timeout must be greater than 0 and at most 300 seconds")
        }
        let config = WKWebViewConfiguration()
        let handler = AssetSchemeHandler(bundle: bundle, scheme: scheme)
        config.setURLSchemeHandler(handler, forURLScheme: scheme)
        self.scheme = scheme
        self.rpcTimeout = rpcTimeout
        webView = WKWebView(frame: .zero, configuration: config)
        super.init()
        config.userContentController.add(self, name: "fylo")
        webView.navigationDelegate = self

        var initializationSucceeded = false
        var initializationFailure: Error = FyloError.failed("FYLO WebView initialization failed")
        defer {
            if !initializationSucceeded { failBridge(initializationFailure) }
        }
        do {
            try await withTaskCancellationHandler(operation: {
                if Task.isCancelled { throw CancellationError() }
                try await withCheckedThrowingContinuation { (cont: CheckedContinuation<Void, Error>) in
                    readyContinuation = cont
                    let timeout = DispatchWorkItem { [weak self] in
                        self?.failBridge(
                            FyloError.failed("FYLO WebView did not become ready within the configured RPC timeout")
                        )
                    }
                    readyTimeout = timeout
                    DispatchQueue.main.asyncAfter(deadline: .now() + rpcTimeout, execute: timeout)
                    webView.load(URLRequest(url: URL(string: "\(scheme)://app/host.html")!))
                }
            }, onCancel: {
                Task { @MainActor [weak self] in
                    self?.failBridge(CancellationError())
                }
            })
            // First real op initializes the local client.
            _ = try await request("open", ["config": [String: Any]()])
            initializationSucceeded = true
        } catch {
            initializationFailure = error
            throw error
        }
    }

    /// Send one RPC to the engine and await its result.
    @discardableResult
    func request(_ method: String, _ args: [String: Any] = [:]) async throws -> Any? {
        guard !isClosed else { throw FyloError.failed("FYLO client is closed") }
        guard pending.count < Self.maxPendingRequests else {
            throw FyloError.failed("too many pending FYLO requests")
        }
        let id = nextId
        nextId += 1
        let payload: [String: Any] = ["id": id, "method": method, "args": args]
        let data = try JSONSerialization.data(withJSONObject: payload)
        guard data.count <= Self.maxBridgeBytes else {
            throw FyloError.failed("FYLO request exceeds the 6 MiB native bridge limit")
        }
        let base64 = data.base64EncodedString()
        return try await withTaskCancellationHandler(operation: {
            if Task.isCancelled { throw CancellationError() }
            return try await withCheckedThrowingContinuation { cont in
                let timeout = DispatchWorkItem { [weak self] in
                    self?.failRequest(
                        id: id,
                        error: FyloError.failed(
                            "FYLO request '\(method)' timed out after the configured RPC timeout"
                        )
                    )
                }
                pending[id] = PendingRequest(continuation: cont, timeout: timeout)
                DispatchQueue.main.asyncAfter(deadline: .now() + rpcTimeout, execute: timeout)
                // base64 body is safe inside the JS string literal.
                webView.evaluateJavaScript("window.__fyloDispatch(\"\(base64)\")") { _, error in
                    if let error {
                        self.failRequest(id: id, error: error)
                    }
                }
            }
        }, onCancel: {
            Task { @MainActor [weak self] in
                self?.failRequest(
                    id: id,
                    error: CancellationError()
                )
            }
        })
    }

    // MARK: Collections
    // `kind` is accepted for cross-client parity; the on-device engine stores documents untyped.
    @discardableResult func createCollection(_ collection: String, _ kind: String = "document") async throws -> Any? {
        try await request("createCollection", ["collection": collection])
    }
    @discardableResult func dropCollection(_ collection: String) async throws -> Any? {
        try await request("dropCollection", ["collection": collection])
    }
    @discardableResult func inspectCollection(_ collection: String) async throws -> Any? {
        try await request("inspectCollection", ["collection": collection])
    }
    @discardableResult func rebuildCollection(_ collection: String) async throws -> Any? {
        try await request("rebuildCollection", ["collection": collection])
    }

    // MARK: Documents
    @discardableResult func putData(_ collection: String, _ data: [String: Any]) async throws -> Any? {
        try await request("putData", ["collection": collection, "data": data])
    }
    @discardableResult func getDoc(_ collection: String, _ id: String) async throws -> Any? {
        try await request("getDoc", ["collection": collection, "id": id])
    }
    @discardableResult func getMeta(_ collection: String, _ id: String) async throws -> Any? {
        try await request("getMeta", ["collection": collection, "id": id])
    }
    @discardableResult func setMeta(_ collection: String, _ id: String, _ meta: [String: Any]) async throws -> Any? {
        try await request("setMeta", ["collection": collection, "id": id, "meta": meta])
    }
    @discardableResult func getLatest(_ collection: String, _ id: String) async throws -> Any? {
        try await request("getLatest", ["collection": collection, "id": id])
    }
    @discardableResult func patchDoc(_ collection: String, _ id: String, _ newDoc: [String: Any]) async throws -> Any? {
        try await request("patchDoc", ["collection": collection, "id": id, "newDoc": newDoc])
    }
    @discardableResult func delDoc(_ collection: String, _ id: String) async throws -> Any? {
        try await request("delDoc", ["collection": collection, "id": id])
    }
    @discardableResult func restoreDoc(_ collection: String, _ id: String) async throws -> Any? {
        try await request("restoreDoc", ["collection": collection, "id": id])
    }

    // MARK: Query
    @discardableResult func findDocs(_ collection: String, _ query: [String: Any]) async throws -> Any? {
        try await request("findDocs", ["collection": collection, "query": query])
    }

    /// Run raw SQL against the local store. Native interpolation is verbatim —
    /// escape/validate untrusted input yourself.
    @discardableResult func sql(_ statement: String) async throws -> Any? {
        try await request("executeSQL", ["sql": statement])
    }

    /// Collection-scoped facade with short method names, so
    /// `try await db.collection("users").put(data)` reads like the browser client.
    func collection(_ name: String) -> FyloCollection {
        FyloCollection(db: self, name: name)
    }

    /// Stop the private WebView and fail every request that has not completed.
    /// Calling `close()` more than once is safe.
    func close() {
        failBridge(FyloError.failed("FYLO client is closed"))
    }

    private func failRequest(id: Int, error: Error) {
        guard let request = pending.removeValue(forKey: id) else { return }
        request.timeout.cancel()
        request.continuation.resume(throwing: error)
    }

    private func failBridge(_ error: Error) {
        guard !isClosed else { return }
        isClosed = true
        isReady = false
        readyTimeout?.cancel()
        readyTimeout = nil
        readyContinuation?.resume(throwing: error)
        readyContinuation = nil
        let requests = Array(pending.values)
        pending.removeAll()
        for request in requests {
            request.timeout.cancel()
            request.continuation.resume(throwing: error)
        }
        webView.stopLoading()
        webView.navigationDelegate = nil
        webView.configuration.userContentController.removeScriptMessageHandler(forName: "fylo")
    }
}

/// A collection-scoped view; methods drop the leading collection argument.
@MainActor
struct FyloCollection {
    let db: Fylo
    let name: String

    @discardableResult func create(_ kind: String = "document") async throws -> Any? {
        try await db.createCollection(name, kind)
    }
    @discardableResult func drop() async throws -> Any? { try await db.dropCollection(name) }
    @discardableResult func inspect() async throws -> Any? { try await db.inspectCollection(name) }
    @discardableResult func rebuild() async throws -> Any? { try await db.rebuildCollection(name) }
    @discardableResult func put(_ data: [String: Any]) async throws -> Any? {
        try await db.putData(name, data)
    }
    @discardableResult func get(_ id: String) async throws -> Any? { try await db.getDoc(name, id) }
    @discardableResult func getMeta(_ id: String) async throws -> Any? { try await db.getMeta(name, id) }
    @discardableResult func setMeta(_ id: String, _ meta: [String: Any]) async throws -> Any? {
        try await db.setMeta(name, id, meta)
    }
    @discardableResult func latest(_ id: String) async throws -> Any? {
        try await db.getLatest(name, id)
    }
    @discardableResult func patch(_ id: String, _ newDoc: [String: Any]) async throws -> Any? {
        try await db.patchDoc(name, id, newDoc)
    }
    @discardableResult func delete(_ id: String) async throws -> Any? { try await db.delDoc(name, id) }
    @discardableResult func restore(_ id: String) async throws -> Any? {
        try await db.restoreDoc(name, id)
    }
    @discardableResult func find(_ query: [String: Any]) async throws -> Any? {
        try await db.findDocs(name, query)
    }
}

// MARK: - WKScriptMessageHandler (JS → native replies)
extension Fylo: WKScriptMessageHandler {
    func userContentController(_ controller: WKUserContentController, didReceive message: WKScriptMessage) {
        guard let text = message.body as? String else {
            if let reply = message.body as? [String: Any], let id = reply["id"] as? Int {
                failCorrelated(id: id, message: "FYLO bridge reply must be a UTF-8 JSON string")
            }
            return
        }
        let recoveredId = Self.recoverId(from: text)
        guard text.utf8.count <= Self.maxBridgeBytes else {
            if let id = recoveredId {
                failCorrelated(id: id, message: "FYLO bridge response exceeds the 6 MiB limit")
            }
            return
        }
        guard let data = text.data(using: .utf8),
              let reply = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
              let id = reply["id"] as? Int
        else {
            if let id = recoveredId {
                failCorrelated(id: id, message: "invalid FYLO bridge response")
            }
            return
        }
        deliver(id: id, reply: reply)
    }

    private static func recoverId(from text: String) -> Int? {
        let prefix = String(text.prefix(1024))
        let pattern = #"^\s*\{\s*"id"\s*:\s*([0-9]{1,15})(?=\s*[,}])"#
        guard let expression = try? NSRegularExpression(pattern: pattern),
              let match = expression.firstMatch(
                  in: prefix,
                  range: NSRange(prefix.startIndex..., in: prefix)
              ),
              let range = Range(match.range(at: 1), in: prefix),
              let id = Int(prefix[range]),
              id >= 0
        else { return nil }
        return id
    }

    private func failCorrelated(id: Int, message: String) {
        let error = FyloError.failed(message)
        if id == 0 {
            readyContinuation?.resume(throwing: error)
            readyContinuation = nil
            return
        }
        failRequest(id: id, error: error)
    }

    private func deliver(id: Int, reply: [String: Any]) {
        if id == 0 { // bridge-ready signal
            guard !isClosed, !isReady else { return }
            isReady = true
            readyTimeout?.cancel()
            readyTimeout = nil
            readyContinuation?.resume()
            readyContinuation = nil
            return
        }
        guard let request = pending.removeValue(forKey: id) else { return }
        request.timeout.cancel()
        if reply["ok"] as? Bool == true {
            request.continuation.resume(returning: reply["result"] ?? NSNull())
        } else {
            let error = reply["error"] as? [String: Any]
            request.continuation.resume(
                throwing: FyloError.failed(error?["message"] as? String ?? "fylo error")
            )
        }
    }
}

// Block top-level navigation away from the one bundled host document. Engine
// module subresources are served by the scheme handler and never need a page
// navigation decision.
extension Fylo: WKNavigationDelegate {
    func webView(
        _ webView: WKWebView,
        decidePolicyFor navigationAction: WKNavigationAction,
        decisionHandler: @escaping @MainActor @Sendable (WKNavigationActionPolicy) -> Void
    ) {
        guard !isClosed else {
            decisionHandler(.cancel)
            return
        }
        guard let url = navigationAction.request.url,
              let components = URLComponents(url: url, resolvingAgainstBaseURL: false),
              url.scheme == scheme,
              url.host == "app",
              components.percentEncodedPath == "/host.html",
              url.user == nil,
              url.password == nil,
              url.port == nil,
              url.query == nil,
              url.fragment == nil
        else {
            decisionHandler(.cancel)
            failBridge(FyloError.failed("FYLO WebView blocked an unexpected navigation"))
            return
        }
        if isReady {
            decisionHandler(.cancel)
            failBridge(FyloError.failed("FYLO WebView unexpectedly navigated after becoming ready"))
            return
        }
        decisionHandler(.allow)
    }

    func webView(
        _ webView: WKWebView,
        didFailProvisionalNavigation navigation: WKNavigation!,
        withError error: Error
    ) {
        failBridge(FyloError.failed("FYLO WebView failed to load: \(error.localizedDescription)"))
    }

    func webView(_ webView: WKWebView, didFail navigation: WKNavigation!, withError error: Error) {
        failBridge(FyloError.failed("FYLO WebView navigation failed: \(error.localizedDescription)"))
    }

    func webViewWebContentProcessDidTerminate(_ webView: WKWebView) {
        failBridge(FyloError.failed("FYLO WebView content process terminated; reopen the client"))
    }
}

// MARK: - Custom-scheme asset handler (secure origin so OPFS is available)
private final class AssetSchemeHandler: NSObject, WKURLSchemeHandler {
    private static let assets: [String: (base: String, ext: String, mime: String)] = [
        "/host.html": ("host", "html", "text/html"),
        "/bridge.js": ("bridge", "js", "text/javascript"),
        "/fylo.mjs": ("fylo", "mjs", "text/javascript")
    ]
    private let bundle: Bundle
    private let scheme: String
    init(bundle: Bundle, scheme: String) {
        self.bundle = bundle
        self.scheme = scheme
    }

    func webView(_ webView: WKWebView, start task: WKURLSchemeTask) {
        guard let url = task.request.url else {
            task.didFailWithError(FyloError.failed("no url"))
            return
        }
        guard let components = URLComponents(url: url, resolvingAgainstBaseURL: false),
              url.scheme == scheme,
              url.host == "app",
              url.user == nil,
              url.password == nil,
              url.port == nil,
              url.query == nil,
              url.fragment == nil,
              let asset = Self.assets[components.percentEncodedPath],
              let fileURL = bundle.url(forResource: asset.base, withExtension: asset.ext),
              let data = try? Data(contentsOf: fileURL) else {
            task.didFailWithError(FyloError.failed("FYLO WebView asset is not allowed"))
            return
        }
        var headers = [
            "Content-Type": "\(asset.mime); charset=utf-8",
            "X-Content-Type-Options": "nosniff",
            "Cache-Control": "no-store"
        ]
        if asset.ext == "html" {
            headers["Content-Security-Policy"] = "default-src 'none'; script-src 'self'; connect-src 'none'; object-src 'none'; base-uri 'none'; form-action 'none'; frame-ancestors 'none'"
        }
        let response = HTTPURLResponse(
            url: url,
            statusCode: 200,
            httpVersion: "HTTP/1.1",
            headerFields: headers
        )!
        task.didReceive(response)
        task.didReceive(data)
        task.didFinish()
    }

    func webView(_ webView: WKWebView, stop task: WKURLSchemeTask) {}
}
